db/db_flush_test.cc

```diff
@@ -78,6 +81,71 @@ TEST_F(DBFlushTest, FlushWhileWritingManifest) {
 #endif  // ROCKSDB_LITE
 }
 
+// We had issue when a flush with wait triggered due to ExportColumnFamily can
+// finish before the result is committed to manifest. This is happening because
+// of a simultaneous flush on the column family. Run in loop to reproduce the
+// failure.
+TEST_F(DBFlushTest, FlushBeforeWritingManifestWithCheckpoint) {
+  Options options;
+  options.disable_auto_compactions = true;
+  options.max_background_flushes = 3;
+  options.max_write_buffer_number = 3;
+  options.create_if_missing = true;
+  env_->SetBackgroundThreads(3, Env::HIGH);
+  options.env = env_;
+  DestroyAndReopen(options);
+
+  SyncPoint::GetInstance()->LoadDependency(
+      {{"DBImpl::FlushMemTable:AfterScheduleNonExportFlush",
+        "FlushJob::WriteLevel0Table:flush_started2"}});
+
+
+  bool processed = false;
+  SyncPoint::GetInstance()->SetCallBack(
+      "FlushJob::WriteLevel0Table:flush_started1", [&](void* /*arg*/) {
+        // Add second write and flush without wait.
+        if (processed) {
+          return;
+        }
+
+        processed = true;
+        FlushOptions no_wait;
+        no_wait.wait = false;
+        no_wait.allow_write_stall = true;
+        ASSERT_OK(Put("foo", "v"));
+        ASSERT_OK(dbfull()->Flush(no_wait));
+      });
+
+  SyncPoint::GetInstance()->SetCallBack(
+      "FlushJob::WriteLevel0Table:add_delay", [&](void* /*arg*/) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(rand() % 100));
+      });
+
+  SyncPoint::GetInstance()->EnableProcessing();
+
+  rocksdb::ExportImportFilesMetaData* cf_sst_files_metadata = nullptr;
+  rocksdb::Checkpoint* checkpoint;
+  auto status = rocksdb::Checkpoint::Create(dbfull(), &checkpoint);
+  ASSERT_EQ(status.ok(), true);
+
+  // Add first write and trigger an ExportColumnFamily which will trigger a
+  // flush.
+  ASSERT_OK(Put("bar", "v"));
+  uint64_t timeSinceEpochMilliseconds =
+      std::chrono::duration_cast<std::chrono::milliseconds>(
+          std::chrono::system_clock::now().time_since_epoch())
+          .count();
+  std::string dir_path =
+      "/tmp/checkpoint1_" + std::to_string(timeSinceEpochMilliseconds);
+  status = checkpoint->ExportColumnFamily(dbfull()->DefaultColumnFamily(),
+                                          dir_path, &cf_sst_files_metadata);
+  ASSERT_EQ(status.ok(), true);
+
+#ifndef ROCKSDB_LITE
+  ASSERT_GT(cf_sst_files_metadata->files.size(), 0);
+#endif  // ROCKSDB_LITE
+}
```

Comment:
At current stage as a temporary thing, if you insist on using checkpoint/export column family, you need to cleanup the checkpoint dir from /tmp so that people can run `gtest-parallel` with your unit test.

Reply:
Done